buck2/rust/BUCK

```diff
@@ -0,0 +1,59 @@
+# Copyright 2022 EngFlow Inc. All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+rust_library(
+    name = "library",
+    # We set these env vars to configure the RE Rust toolchain. Unfortunately the prelude does not provide a way to set these at the rust_toolchain level. We override the HOME location so rustp creates its `.rustp` directory in this location.
+    env = {
+        "RUSTUP_TOOLCHAIN":"/usr/local/rustup/toolchains/1.83.0-x86_64-unknown-linux-gnu",
```

Comment:
I have some questions:
- Why are you specifying these env properties on the target when they are shared by all targets?
- What happens with local execution? This setup only works with a very specific system configuration
- Why are you setting HOME? And why are you setting it to buck-out?
- If you're providing a toolchain for RE anyway, shouldn't all of these things be part of the toolchain instead?

Reply:
Thanks for the review, these are very valid concerns.

> What happens with local execution?
- The way to achieve compatibility with local execution might be to use select statements that are able to only add these vars when running RE. I can attempt to add these to overcome that limitation. 

> Why are you setting HOME? And why are you setting it to buck-out?
- The reason I need to set a different value for `HOME` is that `rustp` uses this env var to determine where to create a `.rustp` directory. I set it to `buck2-out` which avoids the issue of having it in the default home directory and moves it to the same directory where all other outputs of the build are located. 
> If you're providing a toolchain for RE anyway, shouldn't all of these things be part of the toolchain instead?
- Unfortunately the current solution highlights the implementation of `rust_toolchain` in the prelude is not really well tailored to RE. I thought about this for a while and its definitely not convenient. 
- The most convenient long term solution would be to add a `rustc_env` attr to the `rust_toolchain`, similar to the current `rustdoc_env` attr - https://github.com/facebook/buck2/blob/main/prelude/rust/rust_toolchain.bzl#L57 and plumb it through the rules (e.g.,here - https://github.com/facebook/buck2/blob/main/prelude/rust/build.bzl#L372 and in other rust *.bzl files) so that it can be set at the toolchain level.

Concrete next steps: 
- I can try to use select statements to make the env vars be set only for RE
- I can open up an issue in the prelude to add the rustc_env` attr and prepare a PR to show what it would look like
- Submit this PR with the current env set in the rust_* targets and modify it once the upstream change to prelude is in.

Let me know what you think. Thanks for the review!

Reply:
> What happens with local execution?
Addressed by adding select statements and a config for engflow re.